File: src/main/java/ru/nucodelabs/algorithms/charts/MisfitValuesFactory.java
package ru.nucodelabs.algorithms.charts;


import ru.nucodelabs.data.ves.ExperimentalData;
import ru.nucodelabs.data.ves.ModelLayer;

import java.util.List;
import java.util.function.BiFunction;

public interface MisfitValuesFactory extends BiFunction<List<ExperimentalData>, List<ModelLayer>, List<Double>> {

    MisfitValuesFactory DEFAULT_IMPL = new MisfitValuesFactoryNative();

    static MisfitValuesFactory getDefaultMisfitValuesFactory() {
        return DEFAULT_IMPL;
    }

    @Override
    List<Double> apply(List<ExperimentalData> experimentalData, List<ModelLayer> modelData);
}

File: src/main/java/ru/nucodelabs/algorithms/charts/MisfitValuesFactory.kt
package ru.nucodelabs.algorithms.charts

import ru.nucodelabs.algorithms.forward_solver.ForwardSolver
import ru.nucodelabs.data.ves.ExperimentalData
import ru.nucodelabs.data.ves.ModelLayer

interface MisfitValuesFactory {
    /**
     * Returns list of misfits between experimental and theoretical curves
     */
    operator fun invoke(experimentalData: List<ExperimentalData>, modelData: List<ModelLayer>): List<Double>

    companion object Factory {
        @JvmStatic
        fun createDefault(forwardSolver: ForwardSolver): MisfitValuesFactory = MisfitValuesFactory(forwardSolver)
    }
}

fun MisfitValuesFactory(forwardSolver: ForwardSolver): MisfitValuesFactory = MisfitValuesFactoryNative(forwardSolver)